[BUG]kl_ctl大于0时（使用reference model），训练报错

## Checklist

- [√] The error occurs when using our provided Docker image.
- [√] I can consistently reproduce the bug across multiple trials or random seeds.
- [√] If the error causes experiment abortion, I've verified that this error is the root
  cause, not a secondary error caused by peer workers.

## Detailed Information

### Describe the bug

在昇腾环境下，设置kl_ctl > 0, 模型训练报错；相同代码当kl_ctl=0时，可正常运行

### Expected behavior


### Full logs

```
(AReaL) 20260327-07:55:03.823 SyncRPCServer ERROR: Engine method 'compute_logp' failed: create_config:build/CMakeFiles/torch_npu.dir/compiler_depend.ts:130 HCCL function error: hcclCommInitRootInfoConfig(numRanks, &rootInfo, rank, config, &(comm->hcclComm_)), error code is 7
[ERROR] 2026-03-27-07:55:03 (PID:16066, Device:0, RankID:1) ERR02200 DIST call hccl api failed.
[PID: 16066] 2026-03-27-07:55:03.676.007 HCCP_Process_Initialization_Failure(EJ0001): Failed to initialize the HCCP process. Reason: Maybe the last training process is running.
        Solution: Wait for 10s after killing the last training process and try again.
        TraceBack (most recent call last):
        tsd client wait response fail, hostpid:1577678, device response code[1]. unknown device error.[FUNC:WaitRsp][FILE:process_mode_manager.cpp][LINE:300]        Starting a subprocess on the device timed out.
        rtOpenNetService execution failed, reason=open tsd error[FUNC:FuncErrorReason][FILE:error_message_manage.cc][LINE:61]

Traceback (most recent call last):
  File "/workspace/AReaL/areal/infra/rpc/rpc_server.py", line 750, in execute_in_engine_thread
    result = method(*args_bcast, **kwargs_bcast)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/engine/fsdp_engine.py", line 1640, in compute_logp
    return self.actor.compute_logp(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/utils/perf_tracer.py", line 1958, in sync_wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/trainer/ppo/actor.py", line 123, in compute_logp
    return self.engine.forward(
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/api/engine_api.py", line 472, in forward
    return self.forward_batch(input_, output_seqlens, aggregate_fn)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/engine/fsdp_engine.py", line 673, in forward_batch
    mb_list = self._prepare_mb_list(input_).to(self.device)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/engine/fsdp_engine.py", line 1307, in _prepare_mb_list
    mb_list = split_padded_tensor_dict_into_mb_list(input_, self.config.mb_spec)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/utils/data.py", line 541, in split_padded_tensor_dict_into_mb_list
    group_indices = allocate_balanced_mbs_synced(mb_spec, input_lens, group=group)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/utils/data.py", line 264, in allocate_balanced_mbs_synced
    all_n_mbs = all_gather_int(len(group_indices), group=group)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/utils/data.py", line 278, in all_gather_int
    dist.all_gather(out, x, group=group)
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/distributed/c10d_logger.py", line 81, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py", line 3942, in all_gather
    work = group.allgather([tensor_list], [tensor], opts)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: create_config:build/CMakeFiles/torch_npu.dir/compiler_depend.ts:130 HCCL function error: hcclCommInitRootInfoConfig(numRanks, &rootInfo, rank, config, &(comm->hcclComm_)), error code is 7
[ERROR] 2026-03-27-07:55:03 (PID:16066, Device:0, RankID:1) ERR02200 DIST call hccl api failed.
[PID: 16066] 2026-03-27-07:55:03.676.007 HCCP_Process_Initialization_Failure(EJ0001): Failed to initialize the HCCP process. Reason: Maybe the last training process is running.
        Solution: Wait for 10s after killing the last training process and try again.
        TraceBack (most recent call last):
        tsd client wait response fail, hostpid:1577678, device response code[1]. unknown device error.[FUNC:WaitRsp][FILE:process_mode_manager.cpp][LINE:300]        Starting a subprocess on the device timed out.
        rtOpenNetService execution failed, reason=open tsd error[FUNC:FuncErrorReason][FILE:error_message_manage.cc][LINE:61]


(AReaL) 20260327-07:55:03.827 RLTrainer ERROR: Training failed with exception: Failed to call method 'compute_logp' on worker 'ref/1': Unknown error
Traceback (most recent call last):
  File "/workspace/AReaL/examples/multi_turn_math/gsm8k_rl_mt_v1.py", line 146, in main
    trainer.train(
  File "/workspace/AReaL/areal/trainer/rl_trainer.py", line 398, in train
    rollout_batch["ref_logp"] = self.ref.compute_logp(rollout_batch)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/trainer/ppo/actor.py", line 349, in compute_logp
    return self._custom_function_call("compute_logp", *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/infra/controller/train_controller.py", line 307, in _custom_function_call
    results = run_async_task(
              ^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/infra/utils/concurrent.py", line 69, in run_async_task
    return asyncio.run(func(*args, **kwargs))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
  File "/workspace/AReaL/areal/infra/controller/train_controller.py", line 388, in _call_with_dispatched_inputs
    return await asyncio.gather(*tasks)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/infra/scheduler/local.py", line 1411, in async_call_engine
    raise EngineCallError(
areal.infra.scheduler.exceptions.EngineCallError: Failed to call method 'compute_logp' on worker 'ref/1': Unknown error
(AReaL) 20260327-07:55:03.832 StatsLogger INFO: Training completes! Total time elapsed 226.38.
(APIServer pid=16230) INFO 03-27 07:55:09 [loggers.py:257] Engine 000: Avg prompt throughput: 2085.7 tokens/s, Avg generation throughput: 3874.0 tokens/s, Running: 255 reqs, Waiting: 519 reqs, GPU KV cache usage: 6.5%, Prefix cache hit rate: 0.0%
swanlab: 🏠 View project at https://swanlab.cn/@xiaobai/gsm8k-grpo-mt
swanlab: 🚀 View run at 
https://swanlab.cn/@xiaobai/gsm8k-grpo-mt/runs/i9of3puqdl3qlaoz6cf7j
(AReaL) 20260327-07:55:13.571 RolloutController INFO: Stopping callback server...
(AReaL) 20260327-07:55:13.899 LocalScheduler INFO: Removing forked role 'eval-rollout' (managed by parent worker)
(AReaL) 20260327-07:55:13.900 RolloutController INFO: Workers deleted
(AReaL) 20260327-07:55:13.907 RolloutController INFO: Stopping callback server...
(AReaL) 20260327-07:55:14.493 ProcUtils INFO: Sending SIGTERM to process 16230 and 3 children
(Worker pid=16313) INFO 03-27 07:55:14 [multiproc_executor.py:707] Parent process exited, terminating worker
(APIServer pid=16230) INFO 03-27 07:55:15 [launcher.py:110] Shutting down FastAPI HTTP server.
(AReaL) 20260327-07:55:19.495 ProcUtils WARNING: Force killing 3 processes that didn't terminate gracefully
(AReaL) 20260327-07:55:20.497 ProcUtils INFO: Successfully cleaned up process tree for PID 16230
(AReaL) 20260327-07:55:20.499 LocalScheduler INFO: Deleting 1 workers for role 'rollout'
(AReaL) 20260327-07:55:20.508 ProcUtils INFO: Sending SIGTERM to process 16196 and 8 children
(AReaL) 20260327-07:55:22.472 ProcUtils INFO: Successfully cleaned up process tree for PID 16196
(AReaL) 20260327-07:55:22.472 LocalScheduler INFO: Successfully deleted workers for role 'rollout'
(AReaL) 20260327-07:55:22.473 RolloutController INFO: Workers deleted
(AReaL) 20260327-07:55:22.473 TrainController INFO: Destroying TrainController...
(AReaL) 20260327-07:55:22.474 TrainController INFO: Destroying engines on all workers...
(AReaL) 20260327-07:55:22.483 SyncRPCServer ERROR: Engine method 'destroy' failed: ACL stream synchronize failed, error code:507901
Traceback (most recent call last):
  File "/workspace/AReaL/areal/infra/rpc/rpc_server.py", line 678, in execute_in_engine_thread
    raw_args = broadcast_tensor_container(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/AReaL/areal/utils/data.py", line 1072, in broadcast_tensor_container
    dist.broadcast_object_list(metadata, src=src_rank, group=group)
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/distributed/c10d_logger.py", line 81, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py", line 3682, in broadcast_object_list
    *[_object_to_tensor(obj, current_device, group) for obj in object_list]
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py", line 3682, in <listcomp>
    *[_object_to_tensor(obj, current_device, group) for obj in object_list]
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/python3.11.14/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py", line 3076, in _object_to_tensor
    byte_tensor = torch.ByteTensor(byte_storage).to(device)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: ACL stream synchronize failed, error code:507901
```

### Script

The bash script or YAML configuration to run:
```
experiment_name: gsm8k-grpo-mt
trial_name: trial0

seed: 1
enable_offload: false
total_train_epochs: 5
tokenizer_path: ${actor.path}

export_style: concat
agent_run_args:
  max_turns: 2

cluster:
  n_nodes: 1
  n_gpus_per_node: 4
  fileroot: /tmp/areal/experiments
  name_resolve:
    type: nfs
    nfs_record_root: /tmp/areal/name_resolve

allocation_mode: vllm:d1p1t1+d2p1t1

scheduler:
  type: null

rollout:
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  max_concurrent_rollouts: 256
  queue_size: null
  consumer_batch_size: ${train_dataset.batch_size}
  max_head_offpolicyness: 2
  enable_rollout_tracing: false
  scheduling_spec: ${actor.scheduling_spec}
  fileroot: ${cluster.fileroot}
  tokenizer_path: ${tokenizer_path}
  dump_to_file: true

gconfig:
  n_samples: 4
  min_new_tokens: 0
  max_new_tokens: 1024
  greedy: false
  temperature: 1.0

actor:
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  path: /home/Qwen2.5-1.5B-Instruct
  init_from_scratch: false
  disable_dropout: true
  gradient_checkpointing: true
  dtype: bfloat16
  mb_spec:
    max_tokens_per_mb: 10240
  optimizer:
    type: adam
    lr: 1.70e-5
    weight_decay: 0.017
    beta1: 0.9
    beta2: 0.999
    eps: 1e-8
    lr_scheduler_type: constant
    gradient_clipping: 1.0
    warmup_steps_proportion: 0.001
  eps_clip: 0.4
  temperature: ${gconfig.temperature}
  reward_scaling: 10.0
  reward_bias: -0.5
  kl_ctl: 0.1
  kl_estimator: k3
  ppo_n_minibatches: 1
  recompute_logprob: true
  use_decoupled_loss: true
  behave_imp_weight_cap: 5.0
  reward_norm: null
  adv_norm:
    mean_level: batch
    std_level: batch
  weight_update_mode: xccl
  max_new_tokens: ${gconfig.max_new_tokens}
  scheduling_spec:
    - task_type: worker
      port_count: 2
      gpu: 1
      mem: 32
      cmd: python3 -m areal.infra.rpc.rpc_server
      env_vars:
        VLLM_USE_V1: "1"
        VLLM_ALLOW_LONG_MAX_MODEL_LEN: "1"
        TASK_QUEUE_ENABLE: "1"
        HCCL_EXEC_TIMEOUT: "14400"
        HCCL_OP_EXPANSION_MODE: "HOST"
        ACL_DEVICE_SYNC_TIMEOUT: "14400"
        HCCL_EVENT_TIMEOUT: "14500"
        HCCL_ASYNC_ERROR_HANDLING: "0"
        ACL_STREAM_TIMEOUT: "14500000"
        HCCL_CONNECT_TIMEOUT: "7200"
        PYTORCH_NPU_ALLOC_CONF: "expandable_segments:True"

ref:
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  path: ${actor.path}
  init_from_scratch: false
  disable_dropout: true
  dtype: ${actor.dtype}
  mb_spec:
    max_tokens_per_mb: 10240
  optimizer: null
  scheduling_strategy:
    type: colocation
    target: actor
  scheduling_spec: ${actor.scheduling_spec}

# SGLang
sglang:
  model_path: ${actor.path}
  random_seed: ${seed}
  skip_tokenizer_init: true
  dtype: ${actor.dtype}
  max_running_requests: null
  context_length: 32768
  mem_fraction_static: 0.8

vllm:
  model: ${actor.path}
  seed: ${seed}
  skip_tokenizer_init: false
  dtype: ${actor.dtype}
  max_model_len: 32768
  gpu_memory_utilization: 0.9

# datasets
train_dataset:
  batch_size: 256
  shuffle: true
  pin_memory: true
  num_workers: 4
  path: openai/gsm8k
  type: rl
  max_length: 1024

valid_dataset:
  batch_size: 256
  pin_memory: true
  num_workers: 4
  path: openai/gsm8k
  type: rl

# Utilities
saver:
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  fileroot: ${cluster.fileroot}
  freq_epochs: 1
  freq_steps: null
  freq_secs: null

recover:
  mode: disabled
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  fileroot: ${cluster.fileroot}
  freq_epochs: 1
  freq_steps: null
  freq_secs: 3600

evaluator:
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  fileroot: ${cluster.fileroot}
  freq_epochs: 1
  freq_steps: null
  freq_secs: null

stats_logger:
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  fileroot: ${cluster.fileroot}
  swanlab:
    mode: online
    api_key: "xxxx" 
    config: {} 
    logdir: /home/logs

perf_tracer:
  experiment_name: ${experiment_name}
  trial_name: ${trial_name}
  fileroot: ${cluster.fileroot}
  enabled: false
  session_tracer:
    enabled: false
```

--- a/areal/engine/core/distributed.py
+++ b/areal/engine/core/distributed.py
@@ -1,8 +1,10 @@
 # SPDX-License-Identifier: Apache-2.0
 
+import os
 from datetime import timedelta
 
-import torch  # noqa: F401 (used by callers)
+import torch
+import torch.distributed as dist
 
 
 def patch_dist_group_timeout(timeout: timedelta):
@@ -19,6 +21,51 @@ def patch_dist_group_timeout(timeout: timedelta):
 
     if hasattr(distributed_c10d, "default_pg_nccl_timeout"):
         distributed_c10d.default_pg_nccl_timeout = timeout
+
+
+def warmup_process_groups(*groups: dist.ProcessGroup | None) -> None:
+    """Force eager initialization of the collective communicator for each group.
+
+    NCCL/HCCL communicators are created lazily on the first collective call.
+    On Ascend NPU (HCCL), deferring init until a collective runs during
+    training is prone to fail with HCCP process initialization errors
+    (e.g. ``hcclCommInitRootInfoConfig`` error code 7) when multiple
+    colocated engines (for example actor + reference) independently mint
+    overlapping subgroups and trigger their first collective in the middle
+    of training work. Running a small dummy all-reduce at setup time forces
+    the communicator to be initialized while all ranks are aligned and the
+    device is idle, which avoids the race.
+
+    ``None`` groups and duplicates are skipped. No-op on CPU-only platforms
+    or before ``dist.init_process_group``. Safe to call repeatedly;
+    subsequent calls on already-initialized groups are cheap.
+    """
+    # Deferred import to keep this module importable without a platform
+    # (e.g. during lightweight tooling or unit tests).
+    from areal.infra.platforms import current_platform
+
+    if not dist.is_initialized() or current_platform.device_type == "cpu":
+        return
+
+    seen: set[int] = set()
+    unique_groups: list[dist.ProcessGroup] = []
+    for group in groups:
+        if group is None:
+            continue
+        key = id(group)
+        if key in seen:
+            continue
+        seen.add(key)
+        unique_groups.append(group)
+    if not unique_groups:
+        return
+
+    local_rank = int(os.environ["LOCAL_RANK"])
+    current_platform.set_device(local_rank)
+    device = torch.device(current_platform.device_type, local_rank)
+    tensor = torch.zeros(1, device=device)
+    for group in unique_groups:
+        dist.all_reduce(tensor, group=group)
 
 
 # Copy from pytorch and OpenRLHF to allow creating multiple main groups.

--- a/areal/engine/core/distributed.py
+++ b/areal/engine/core/distributed.py
@@ -47,21 +47,21 @@ def warmup_process_groups(*groups: dist.ProcessGroup | None) -> None:
     if not dist.is_initialized() or current_platform.device_type == "cpu":
         return
 
-    seen: set[int] = set()
-    unique_groups: list[dist.ProcessGroup] = []
-    for group in groups:
-        if group is None:
-            continue
-        key = id(group)
-        if key in seen:
-            continue
-        seen.add(key)
-        unique_groups.append(group)
+    # Preserve argument order while dropping None and duplicates.
+    unique_groups = list(dict.fromkeys(g for g in groups if g is not None))
     if not unique_groups:
         return
 
-    local_rank = int(os.environ["LOCAL_RANK"])
-    current_platform.set_device(local_rank)
+    # Prefer LOCAL_RANK (set by torchrun/most launchers); fall back to the
+    # device the caller has already configured via ``set_device``. This keeps
+    # the helper usable under custom launchers that don't export LOCAL_RANK.
+    local_rank_env = os.environ.get("LOCAL_RANK")
+    if local_rank_env is not None:
+        local_rank = int(local_rank_env)
+        current_platform.set_device(local_rank)
+    else:
+        local_rank = current_platform.current_device()
+
     device = torch.device(current_platform.device_type, local_rank)
     tensor = torch.zeros(1, device=device)
     for group in unique_groups:

--- a/areal/engine/fsdp_engine.py
+++ b/areal/engine/fsdp_engine.py
@@ -64,6 +64,7 @@ from areal.engine.core import (
 from areal.engine.core.distributed import (
     init_custom_process_group,
     patch_dist_group_timeout,
+    warmup_process_groups,
 )
 from areal.engine.core.model import (
     disable_dropout_in_model,
@@ -266,6 +267,10 @@ class FSDPEngine(TrainEngine):
         self.dp_rank = dist.get_rank(self.dp_group)
 
         self.logger.info(f"Data parallel head {self.dp_head} and rank {self.dp_rank}")
+
+        # Eagerly initialize HCCL/NCCL communicators for the subgroups so
+        # that lazy init doesn't race with colocated engines (issue #1099).
+        warmup_process_groups(self.dp_group, self.sp_group, self.mp_group)
 
     def initialize(self, addr: str | None, ft_spec: FinetuneSpec, *args, **kwargs):
         # Initialize distributed enviroments and load model.

--- a/areal/engine/megatron_engine.py
+++ b/areal/engine/megatron_engine.py
@@ -51,7 +51,10 @@ from areal.engine.core import (
     compute_total_loss_weight,
     reorder_and_pad_outputs,
 )
-from areal.engine.core.distributed import init_custom_process_group
+from areal.engine.core.distributed import (
+    init_custom_process_group,
+    warmup_process_groups,
+)
 from areal.engine.core.model import disable_dropout_in_model
 from areal.engine.megatron_utils.checkpointer import MegatronCheckpointManager
 from areal.engine.megatron_utils.deterministic import set_deterministic_algorithms
@@ -218,6 +221,13 @@ class MegatronEngine(TrainEngine):
             timeout=DIST_GROUP_DEFAULT_TIMEOUT, backend="gloo"
         )
         self.process_group_initialized = True
+
+        # Eagerly initialize HCCL/NCCL communicators for the subgroups so
+        # that lazy init doesn't race with colocated engines (issue #1099).
+        warmup_process_groups(
+            self._context_and_model_parallel_group,
+            mpu.get_data_parallel_group(),
+        )
 
     def _apply_megatron_bridge_lora(self) -> None:
         assert self.model is not None, "Model must be initialized before applying LoRA."

--- a/areal/experimental/engine/archon_engine.py
+++ b/areal/experimental/engine/archon_engine.py
@@ -35,7 +35,10 @@ from areal.api import (
 )
 from areal.api.cli_args import MicroBatchSpec
 from areal.api.io_struct import DeviceRuntimeInfo
-from areal.engine.core.distributed import patch_dist_group_timeout
+from areal.engine.core.distributed import (
+    patch_dist_group_timeout,
+    warmup_process_groups,
+)
 from areal.engine.core.train_engine import (
     aggregate_eval_losses,
     compute_total_loss_weight,
@@ -282,6 +285,15 @@ class ArchonEngine(TrainEngine):
             f"pp={self.parallel_dims.pp}, dp_shard={self.parallel_dims.dp_shard}, "
             f"tp={self.parallel_dims.tp}, cp={self.parallel_dims.cp} (Ulysses SP), "
             f"ep={self.parallel_dims.ep}, etp={self.parallel_dims.etp}"
+        )
+
+        # Eagerly initialize HCCL/NCCL communicators for the subgroups so
+        # that lazy init doesn't race with colocated engines (issue #1099).
+        warmup_process_groups(
+            self.parallel_dims.world_mesh["dp"].get_group(),
+            self._pp_cp_tp_group,
+            self._tp_group,
+            self._cp_group,
         )
 
     def initialize(self, addr: str | None, ft_spec: FinetuneSpec, *args, **kwargs):

--- a/tests/test_warmup_process_groups.py
+++ b/tests/test_warmup_process_groups.py
@@ -61,10 +61,29 @@ def test_dedupes_repeated_groups(mock_all_reduce, monkeypatch):
         patch("torch.zeros") as zeros,
     ):
         platform.device_type = "cuda"
-        platform.set_device = lambda _: None
         zeros.return_value = object()
         warmup_process_groups(group, group, None, group)
 
     assert mock_all_reduce.call_count == 1
+    platform.set_device.assert_called_once_with(0)
     kwargs = mock_all_reduce.call_args.kwargs
     assert kwargs["group"] is group
+
+
+def test_falls_back_to_current_device_without_local_rank(mock_all_reduce, monkeypatch):
+    """When LOCAL_RANK is unset, the caller's current device is used."""
+    monkeypatch.delenv("LOCAL_RANK", raising=False)
+    group = object()
+    with (
+        patch.object(dist, "is_initialized", return_value=True),
+        patch("areal.infra.platforms.current_platform") as platform,
+        patch("torch.zeros") as zeros,
+    ):
+        platform.device_type = "cuda"
+        platform.current_device.return_value = 3
+        zeros.return_value = object()
+        warmup_process_groups(group)
+
+    platform.current_device.assert_called_once_with()
+    platform.set_device.assert_not_called()
+    assert mock_all_reduce.call_count == 1

--- a/tests/test_warmup_process_groups.py
+++ b/tests/test_warmup_process_groups.py
@@ -1,0 +1,70 @@
+"""Unit tests for ``warmup_process_groups``.
+
+The collective warmup itself needs a distributed environment, but the
+no-op short-circuits can be exercised with plain unit tests so we catch
+regressions in the guard logic without requiring a GPU.
+"""
+
+from __future__ import annotations
+
+from unittest.mock import patch
+
+import pytest
+import torch.distributed as dist
+
+from areal.engine.core.distributed import warmup_process_groups
+
+
+@pytest.fixture
+def mock_all_reduce():
+    with patch.object(dist, "all_reduce") as m:
+        yield m
+
+
+def test_noop_when_dist_uninitialized(mock_all_reduce):
+    with patch.object(dist, "is_initialized", return_value=False):
+        warmup_process_groups(object(), object())
+    mock_all_reduce.assert_not_called()
+
+
+def test_noop_on_cpu_platform(mock_all_reduce):
+    with (
+        patch.object(dist, "is_initialized", return_value=True),
+        patch("areal.infra.platforms.current_platform") as platform,
+    ):
+        platform.device_type = "cpu"
+        warmup_process_groups(object(), object())
+    mock_all_reduce.assert_not_called()
+
+
+def test_noop_when_all_groups_are_none(mock_all_reduce):
+    with (
+        patch.object(dist, "is_initialized", return_value=True),
+        patch("areal.infra.platforms.current_platform") as platform,
+    ):
+        platform.device_type = "cuda"
+        warmup_process_groups(None, None)
+    mock_all_reduce.assert_not_called()
+
+
+def test_noop_with_no_arguments(mock_all_reduce):
+    warmup_process_groups()
+    mock_all_reduce.assert_not_called()
+
+
+def test_dedupes_repeated_groups(mock_all_reduce, monkeypatch):
+    monkeypatch.setenv("LOCAL_RANK", "0")
+    group = object()
+    with (
+        patch.object(dist, "is_initialized", return_value=True),
+        patch("areal.infra.platforms.current_platform") as platform,
+        patch("torch.zeros") as zeros,
+    ):
+        platform.device_type = "cuda"
+        platform.set_device = lambda _: None
+        zeros.return_value = object()
+        warmup_process_groups(group, group, None, group)
+
+    assert mock_all_reduce.call_count == 1
+    kwargs = mock_all_reduce.call_args.kwargs
+    assert kwargs["group"] is group

